RiotSwiftUI/Modules/Authentication/Common/Service/MatrixSDK/RegistrationFlowHandling.swift

```diff
@@ -0,0 +1,77 @@
+// 
+// Copyright 2022 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+@available(iOS 14.0, *)
+/// A protocol with a default implementation that allows a coordinator to execute and handle registration flow steps.
+protocol RegistrationFlowHandling {
+    var authenticationService: AuthenticationService { get }
+    var registrationWizard: RegistrationWizard? { get }
+    @MainActor var completion: ((AuthenticationRegistrationCoordinatorResult) -> Void)? { get }
+    
+    /// Executes a registration step using the `RegistrationWizard` to complete any additional steps automatically.
+    @MainActor func executeRegistrationStep(step: @escaping (RegistrationWizard) async throws -> RegistrationResult) -> Task<Void, Error>
+}
+
+@available(iOS 14.0, *)
+@MainActor extension RegistrationFlowHandling {
+    func executeRegistrationStep(step: @escaping (RegistrationWizard) async throws -> RegistrationResult) -> Task<Void, Error> {
+        return Task {
+            guard let registrationWizard = registrationWizard else {
+                MXLog.failure("[RegistrationFlowHandling] executeRegistrationStep: The registration wizard is nil.")
+                return
+            }
+            
+            do {
+                let result = try await step(registrationWizard)
+                
+                guard !Task.isCancelled else { return }
+                
+                switch result {
+                case .success(let mxSession):
+                    completion?(.sessionCreated(session: mxSession, isAccountCreated: true))
```

Comment:
It is a bit weird to use all the new async code and yet still report back the result via a callback in one branch and via Void values in other. Couldn't the `RegistrationWizard` instead be a part of the return value, e.g. `Task<RegistrationWizard?, Error>`?

Reply:
This method was the one part of the implementation that I brought over which I really wasn't keen on. I've removed it completely and given the responsibility of handling the dummy stage to the `RegistrationWizard` instead.